botorch/acquisition/multi_objective/predictive_entropy_search.py

```diff
@@ -0,0 +1,1179 @@
+        omega_f_nat_cov: A `batch_shape x M x (N + P) x P x 2 x 2`-dim Tensor
+            containing the omega natural covariance factors for the objective matrix.
+        N: The number of design points.
+        P: The number of Pareto optimal points.
+
+    Returns:
+        A two-element tuple containing
+
+        - nat_mean_f: A `batch_shape x M x (N + P)`-dim Tensor containing the updated
+            natural mean matrix for the objectives.
+        - nat_cov_f: A `batch_shape x M x (N + P) x (N + P)`-dim Tensor containing
+            the updated natural predictive covariance matrix for the objectives.
+    """
+
+    # `batch_shape x M x (N + P)`
+    nat_mean_f = pred_f_nat_mean.clone()
+    # `batch_shape x M x (N + P) x (N + P)
+    nat_cov_f = pred_f_nat_cov.clone()
+
+    ################################################################################
+    # UPDATE THE OBJECTIVES
+    ################################################################################
+    # remove Pareto diagonal
+    # zero out the diagonal
+    omega_f_nat_mean[..., range(N, N + P), range(P), :] = 0
+    omega_f_nat_cov[..., range(N, N + P), range(P), :, :] = 0
+
+    # `batch_shape x M x (N + P)`
+    # sum over the pareto dim
+    nat_mean_f = nat_mean_f + omega_f_nat_mean[..., 0].sum(dim=-1)
+    # `batch_shape x M x P`
+    # sum over the data dim
+    nat_mean_f[..., N:] = nat_mean_f[..., N:] + omega_f_nat_mean[..., 1].sum(dim=-2)
+
+    # `batch_shape x M x (N + P)`
+    nat_cov_f[..., range(N + P), range(N + P)] = nat_cov_f[
+        ..., range(N + P), range(N + P)
+    ] + omega_f_nat_cov[..., 0, 0].sum(dim=-1)
+    # `batch_shape x M x P`
+    nat_cov_f[..., range(N, N + P), range(N, N + P)] = nat_cov_f[
+        ..., range(N, N + P), range(N, N + P)
+    ] + omega_f_nat_cov[..., 1, 1].sum(dim=-2)
+
+    for p in range(P):
+        # `batch_shape x M x (N + P)`
+        nat_cov_f[..., range(N + P), N + p] = (
+            nat_cov_f[..., range(N + P), N + p] + omega_f_nat_cov[..., p, 0, 1]
+        )
+
+        # `batch_shape x M x (N + P)`
+        nat_cov_f[..., N + p, range(N + P)] = (
+            nat_cov_f[..., N + p, range(N + P)] + omega_f_nat_cov[..., p, 1, 0]
+        )
+
+    return nat_mean_f, nat_cov_f
+
+
+def _damped_update(
+    old_factor: Tensor,
+    new_factor: Tensor,
+    damping_factor: Tensor,
+) -> Tensor:
+    r"""Computes the damped updated for natural factor.
+
+    Args:
+        old_factor: A `batch_shape x param_shape`-dim Tensor containing the old
+            natural factor.
+        new_factor: A `batch_shape x param_shape`-dim Tensor containing the new
+            natural factor.
+        damping_factor: A `batch_shape`-dim Tensor containing the damping factor.
+
+    Returns:
+        A `batch_shape x param_shape`-dim Tensor containing the updated natural
+        factor.
+    """
+    bs = damping_factor.shape
+    fs = old_factor.shape
+
+    df = damping_factor
+    for i in range(len(fs[len(bs) :])):
```

Comment:
```suggestion
    for _ in range(len(fs[len(bs) :])):
```
Addressing a complaint from an internal linter